### PR TITLE
Add `unittest.runner._WritelnDecorator`

### PR DESCRIPTION
**Pull request by Avasam**
Used by pywin32
- https://github.com/mhammond/pywin32/blob/377a3efb3fd01602555066150e3c76e74fa5aa4f/com/win32com/test/testall.py#L303-L311
- https://github.com/mhammond/pywin32/blob/377a3efb3fd01602555066150e3c76e74fa5aa4f/win32/Lib/pywin32_testutil.py#L257-L267

~~Granted I'd still have to explicitly set the stream type for custom subclasses, or cast the type of the stream when I know it ahead of time.~~ (Edit: TypeVar defaults to the rescue.) At that point I'd rather just write `self.stream.write(arg or "\n")` (even better in pywin32 the string is never empty, so just `self.stream.write("my string")`

### CODE REVIEW

File: stdlib/unittest/runner.pyi [hunk 1]
```diff
@@ -2,13 +2,21 @@ import sys
...
-from _typeshed import Incomplete
 from collections.abc import Callable, Iterable
-from typing import TextIO
-from typing_extensions import TypeAlias
+from typing import Any, TextIO
+from typing_extensions import Never, TypeAlias
 
 _ResultClassType: TypeAlias = Callable[[TextIO, bool, int], unittest.result.TestResult]
 
+# Note: doesn't actually inherit TextIO, but re-exposes all methods of the stream passed to __init__
```
Comment (srittau) [inline stdlib/unittest/runner.pyi:11]: I'm not sure we should even use `TextIO` as annotation for `stream`. This annotation most likely predates the introduction of protocols (I didn't check) and we should use a tight protocol instead. `_WritelnDecorator` could then either just add the methods of that protocol or we could make `TextTestRunner` generic over the stream, although I'm not sure that that's worth it.

Comment (Avasam) [reply stdlib/unittest/runner.pyi:11]: > we could make `TextTestRunner` generic over the stream, although I'm not sure that that's worth it.

I don't think there's any way to fill a class attributes from generics right?
``py
_T= TypeVar("_T", bound=type)
class _WritelnDecorator(_T): ...
```

The closest I can do is copy all the methods from TextIO

File: stdlib/unittest/runner.pyi [hunk 2]
```diff
@@ -2,36 +2,48 @@ import sys
...
-from typing import TextIO
-from typing_extensions import TypeAlias
+from typing import Any, Generic, Protocol, TextIO, TypeVar, type_check_only
+from typing_extensions import Never, TypeAlias
 
-_ResultClassType: TypeAlias = Callable[[TextIO, bool, int], unittest.result.TestResult]
+# Build a Protocol from TextIO's methods.
+# `TextTestRunner.stream` defaults to `sys.stderr` which is typed as `TextIO`
+@type_check_only
+class _TextIOLike(TextIO, Protocol): ...  # type: ignore[misc] # pyright: ignore[reportGeneralTypeIssues]
```
Comment (srittau) [inline stdlib/unittest/runner.pyi:12]: I'd just use the following, as that is what is actually used by the decorator:

```suggestion
@type_check_only
class _TextIOLike(Protocol):
    def write(self, s: str) -> Unused: ...
```

Comment (JelleZijlstra) [reply stdlib/unittest/runner.pyi:12]: Agree. Also, inheriting from both a concrete class and a Protocol isn't meaningful in the type system. We can type ignore it away, but that doesn't make it meaningful.

**Comment by Avasam** [comment]
I think I got it. And still in a way that can be fully inferred by pywin32
image

**Review by srittau — APPROVED** [review]

**Pull request merged**